### PR TITLE
fix(compiler): mutable lifted object may be modified inflight

### LINKED ISSUE #3069: prevent reassigning and mutating lifted objects
Following https://github.com/winglang/wing/pull/3064

It is now possible to capture reassignable/mutable objects
We need to prevent these captured object from being reassigned/modified in inflight

### PR DESCRIPTION
**Pull request by yoav-steinberg**
fixes #3069
The type checker will now modify the type of preflight expressions that are being used inflight to a non-mutable version of the same type: `MutArray` -> `Array`, `MutJson` -> `Json`...

Note that since we change the type of the expression we don't get descriptive error telling us that we're trying to mutate a lifted mutable object, instead we just get the same error we'd expect if the lifted object was non-mutable to begin with. This isn't ideal but it makes the fix simpler. Let me know what you think...

## Checklist

- [x] Title matches [Winglang's style guide](https://www.winglang.io/contributing/start-here/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [x] Tests added (always)
- [x] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Wing Cloud Contribution License](https://github.com/winglang/wing/blob/main/CONTRIBUTION_LICENSE.md)*.

### CODE REVIEW

File: libs/wingc/src/type_check.rs [hunk 1]
```diff
@@ -5599,6 +5604,29 @@ impl<'a> TypeChecker<'a> {
 			.map(|_| base_udt)
 	}
 
+	fn make_immutable(&mut self, type_: TypeRef) -> TypeRef {
+		match *type_ {
+			Type::MutArray(inner) => {
```
Comment (eladb) [inline libs/wingc/src/type_check.rs:5609]: what about `MutJson`?

**Review by eladb — COMMENTED** [review]
Wow that was easier than expected...

File: libs/wingc/src/type_check.rs (hunk 1 above)
Comment (yoav-steinberg) [reply libs/wingc/src/type_check.rs:5609]: What about it? It's handled 4 lines down. Am I missing something?

Comment (eladb) [reply libs/wingc/src/type_check.rs:5609]: No, I've missed it. Sorry!

**Review by eladb — COMMENTED** [review]
Maybe add to docs?

File: examples/tests/invalid/un_mut_lifted_objects.test.w [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+inflight() => {
+  // Same code as above should be an error in inflight
+  ar?.push(2); // Error: push doesn't exist in Array
+  j.set("a", 3); // Error: set doesn't exist in Json
+  st.add(4); // Error: add doesn't exist in Set
+  mp.set("a", 3); // Error: set doesn't exist in Map
+  opt_ar?.push(2); // Error: push doesn't exist in Array
+  recursive_ar.push(MutArray<num>[2]); // Error: push doesn't exist in Array
+  recursive_ar.at(0).push(3); // Error: push doesn't exist in Array
+};
```
Comment (Chriscbr) [inline examples/tests/invalid/un_mut_lifted_objects.test.w:29]: Maybe double check this works with the new array assignment syntax:

```
ar[0] = 5; // Error: Cannot update elements of an immutable Array
```

**Review by Chriscbr — APPROVED** [review]
Slick 🚢

**Pull request merged**